### PR TITLE
Allow per-module error codes

### LINKED ISSUE #9440: Add support for per-module enabling/disabling of error codes
**Feature**

#9172 implemented global error code disabling/enabling as discussed in #8975. It seems like it was always the intention to consider per-module configuration (https://github.com/python/mypy/issues/8975#issuecomment-641167533), but we don't currently have an open issue for it.

**Pitch**

My personal use case is for tests. Typing tests is often just a chore, but it's nice to have whatever type safety you can get for free. I'd want to be able to do:
```
[mypy]
strict = True

[mypy-tests.*]
allow_untyped_defs = True
allow_untyped_calls = True
disable_error_codes = ["var-annotated"]
```
(For this use case, I think there's actually a fair argument to be made that we should suppress var-annotated aka "needs type annotation" errors when checking untyped defs. If this strikes you as reasonable, I can open another issue. Edit: see #8558 for this)

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #9440

This is a bit non-trivial because I decided to make per-module code act as overrides over main section error codes. This looks more natural no me, rather that putting an adjusted list in each section.

I also fix the inline `# mypy: ...` comment error codes, that are currently just ignored. The logic is naturally like this:
* Command line and/or config main section set global codes
* Config sections _adjust_ them per glob/module
* Inline comments adjust them again

So one can e.g. enable code globally, disable it for all tests in config, and then re-enable locally by an inline comment.

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
Btw, this means that inline mypy comments for error codes will start to act as overrides after this PR. Currently they just completely replace the global config, or command line flags.

**Comment by ilevkivskyi** [comment]
Hm, actually I think currently inline mypy comments may not work at all, because they update e.g. `enable_error_code`, but this one is not used for anything. The `Errors` class uses `disabled_error_codes` set, that is currently computed just once in `main.py`.
This PR makes it a bit better (we at least correctly set options for state). But they are still not used for anything, as options for `Errors` are passed just once in `build.py`.

Let me try also fixing inline comments here.

**Comment by ilevkivskyi** [comment]
OK, so I now also fixed the inline error codes. The logic is naturally like this:
* Command line and/or config main section set global codes
* Config sections _adjust_ them per glob/module
* Inline comments further _adjust_ them

So one can e.g. enable code globally, disable it for all tests in config, and then re-enable in a specific test by an inline comment.

**Comment by ilevkivskyi** [comment]
@hauntsaninja This is ready for review now.

File: mypy/errors.py [hunk 1]
```diff
@@ -313,7 +311,13 @@ def simplify_path(self, file: str) -> str:
             file = os.path.normpath(file)
             return remove_path_prefix(file, self.ignore_prefix)
 
-    def set_file(self, file: str, module: str | None, scope: Scope | None = None) -> None:
+    def set_file(
+        self,
+        file: str,
+        module: str | None,
+        scope: Scope | None = None,
+        options: Options | None = None,
```
Comment (hauntsaninja) [inline mypy/errors.py:319]: nit: Should this have a default value? Having it be required would make sure we get all call sites

**Review by hauntsaninja — APPROVED** [review]
Awesome, just one nit!

File: mypy/errors.py (hunk 1 above)
Comment (ilevkivskyi) [reply mypy/errors.py:319]: Yeah, OK, makes sense.

**Review by JukkaL — APPROVED** [review]
Excellent, this is a very useful feature!

**Pull request merged**